core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -553,12 +562,13 @@ private[spark] class ExecutorAllocationManager(
         }
 
         // If this is the last pending task, mark the scheduler queue as empty
-        stageIdToTaskIndices.getOrElseUpdate(stageId, new mutable.HashSet[Int]) += taskIndex
+        stageIdToTaskIndices.getOrElseUpdate(stageId, new mutable.HashSet[String]) += (taskIndex + "." + attemptId)
```

Comment:
if use taskIndex , when the new task (attempt id is not 0) is starting, the  stageIdToTaskIndices will not update, because the taskIndex is the same with the before failed task.